File: services/nuldoc/nuldoc-src/components/TableOfContents.ts
import { TocEntry, TocRoot } from "../markdown/document.ts";
import { elem, Element } from "../dom.ts";

type Props = {
  toc: TocRoot;
};

export default function TableOfContents({ toc }: Props): Element {
  return elem(
    "nav",
    { class: "toc" },
    elem("h2", {}, "目次"),
    elem(
      "ul",
      {},
      ...toc.entries.map((entry) => TocEntryComponent({ entry })),
    ),
  );
}

function TocEntryComponent({ entry }: { entry: TocEntry }): Element {
  return elem(
    "li",
    {},
    elem("a", { href: `#${entry.id}` }, entry.text),
    entry.children.length > 0
      ? elem(
        "ul",
        {},
        ...entry.children.map((child) => TocEntryComponent({ entry: child })),
      )
      : null,
  );
}

File: services/nuldoc/lib/nuldoc/components/table_of_contents.rb
module Nuldoc
  module Components
    class TableOfContents < DOM::HTMLBuilder
      def initialize(toc:)
        super()
        @toc = toc
      end

      def build
        nav(class: 'toc') do
          h2 { text '目次' }
          ul { @toc.items.each { |entry| toc_entry_component(entry) } }
        end
      end

      private

      def toc_entry_component(entry)
        li do
          a(href: "##{entry.id}") { text entry.text }
          ul { entry.children.each { |c| toc_entry_component(c) } } if entry.children.length.positive?
        end
      end
    end
  end
end
